syntax = "proto2";

package action_message;


enum object_type
{
    null_object_type                                          = 0;
    account_object_type                                       = 1;
    account_sequence_object_type                              = 2;
    permission_object_type                                    = 3;
    permission_usage_object_type                              = 4;
    permission_link_object_type                               = 5;
    UNUSED_action_code_object_type                            = 6;
    key_value_object_type                                     = 7;
    index64_object_type                                       = 8;
    index128_object_type                                      = 9;
    index256_object_type                                      = 10;
    index_double_object_type                                  = 11;
    index_long_double_object_type                             = 12;
    global_property_object_type                               = 13;
    dynamic_global_property_object_type                       = 14;
    block_summary_object_type                                 = 15;
    transaction_object_type                                   = 16;
    generated_transaction_object_type                         = 17;
    producer_object_type                                      = 18;
    UNUSED_chain_property_object_type                         = 19;
    account_control_history_object_type                       = 20;     ///< Defined by history_plugin
    UNUSED_account_transaction_history_object_type            = 21;
    UNUSED_transaction_history_object_type                    = 22;
    public_key_history_object_type                            = 23;          ///< Defined by history_plugin
    UNUSED_balance_object_type                                = 24;
    UNUSED_staked_balance_object_type                         = 25;
    UNUSED_producer_votes_object_type                         = 26;
    UNUSED_producer_schedule_object_type                      = 27;
    UNUSED_proxy_vote_object_type                             = 28;
    UNUSED_scope_sequence_object_type                         = 29;
    table_id_object_type                                      = 30;
    resource_limits_object_type                               = 31;
    resource_usage_object_type                                = 32;
    resource_limits_state_object_type                         = 33;
    resource_limits_config_object_type                        = 34;
    account_history_object_type                               = 35;              ///< Defined by history_plugin
    action_history_object_type                                = 36;               ///< Defined by history_plugin
    reversible_block_object_type                              = 37;
    signed_block_object_type                                  = 40;
    block_state_object_type                                   = 41;
}

enum message_id {
    account_create_request                                   = 1000;
    account_create_response                                  = 1001;
    account_guery_request                                    = 1002;
    account_query_response                                   = 1003;
    account_update_request                                   = 1004;
    account_update_response                                  = 1005;
    account_remove_request                                   = 1006;
    account_remove_response                                  = 1007;
    account_sequence_create_request                          = 1008;
    account_sequence_create_response                         = 1009;
    account_sequence_query_request                           = 1010;
    account_sequence_query_response                          = 1011;
    account_sequence_update_request                          = 1012;
    account_sequence_update_response                         = 1013;
    account_sequence_remove_request                          = 1014;
    account_sequence_remove_response                         = 1015;
    permission_create_request                                = 1016;
    permission_create_response                               = 1017;
    permission_query_request                                 = 1018;
    permission_query_response                                = 1019;
    permission_update_request                                = 1020;
    permission_update_response                               = 1021;
    permission_remove_request                                = 1022;
    permission_remove_response                               = 1023;
    permission_usage_create_request                          = 1024;
    permission_usage_create_response                         = 1025;
    permission_usage_query_request                           = 1026;
    permission_usage_query_response                          = 1027;
    permission_usage_update_request                          = 1028;
    permission_usage_update_response                         = 1029;
    permission_usage_remove_request                          = 1030;
    permission_usage_remove_response                         = 1031;
    permission_link_create_request                           = 1032;
    permission_link_create_response                          = 1033;
    permission_link_query_request                            = 1034;
    permission_link_query_response                           = 1035;
    permission_link_update_request                           = 1036;
    permission_link_update_response                          = 1037;
    permission_link_remove_request                           = 1038;
    permission_link_remove_response                          = 1040;
    key_value_create_request                                 = 1041;
    key_value_create_response                                = 1042;
    key_value_query_request                                  = 1043;
    key_value_query_response                                 = 1044;
    key_value_update_request                                 = 1045;
    key_value_update_response                                = 1046;
    key_value_remove_request                                 = 1047;
    key_value_remove_response                                = 1048;
    index64_create_request                                   = 1049;
    index64_create_response                                  = 1050;
    index64_query_request                                    = 1051;
    index64_query_response                                   = 1052;
    index64_update_request                                   = 1053;
    index64_update_response                                  = 1054;
    index64_remove_request                                   = 1055;
    index64_remove_response                                  = 1056;
    index128_create_request                                  = 1057;
    index128_create_response                                 = 1058;
    index128_query_request                                   = 1060;
    index128_query_response                                  = 1061;
    index128_update_request                                  = 1062;
    index128_update_response                                 = 1063;
    index128_remove_request                                  = 1064;
    index128_remove_response                                 = 1065;
    index256_create_request                                  = 1066;
    index256_create_response                                 = 1067;
    index256_query_request                                   = 1068;
    index256_query_response                                  = 1069;
    index256_update_request                                  = 1070;
    index256_update_response                                 = 1071;
    index256_remove_request                                  = 1072;
    index256_remove_response                                 = 1073;
    index_double_create_request                              = 1074;
    index_double_create_response                             = 1075;
    index_double_query_request                               = 1076;
    index_double_query_response                              = 1077;
    index_double_update_request                              = 1078;
    index_double_update_response                             = 1079;
    index_double_remove_request                              = 1080;
    index_double_remove_response                             = 1081;
    index_long_double_create_request                         = 1082;
    index_long_double_create_response                        = 1083;
    index_long_double_query_request                          = 1084;
    index_long_double_query_response                         = 1085;
    index_long_double_update_request                         = 1086;
    index_long_double_update_response                        = 1087;
    index_long_double_remove_request                         = 1088;
    index_long_double_remove_response                        = 1089;
    global_property_create_request                           = 1090;
    global_property_create_response                          = 1091;
    global_property_query_request                            = 1092;
    global_property_query_response                           = 1093;
    global_property_update_request                           = 1094;
    global_property_update_response                          = 1095;
    global_property_remove_request                           = 1096;
    global_property_remove_response                          = 1097;
    dynamic_global_property_create_request                   = 1098;
    dynamic_global_property_create_response                  = 1099;
    dynamic_global_property_query_request                    = 1100;
    dynamic_global_property_query_response                   = 1101;
    dynamic_global_property_update_request                   = 1102;
    dynamic_global_property_update_response                  = 1103;
    dynamic_global_property_remove_request                   = 1104;
    dynamic_global_property_remove_response                  = 1105;
    block_summary_create_request                             = 1106;
    block_summary_create_response                            = 1107;
    block_summary_query_request                              = 1108;
    block_summary_query_response                             = 1109;
    block_summary_update_request                             = 1110;
    block_summary_update_response                            = 1111;
    block_summary_remove_request                             = 1112;
    block_summary_remove_response                            = 1113;
    transaction_create_request                               = 1114;
    transaction_create_response                              = 1115;
    transaction_query_request                                = 1116;
    transaction_query_response                               = 1117;
    transaction_update_request                               = 1118;
    transaction_update_response                              = 1119;
    transaction_remove_request                               = 1120;
    transaction_remove_response                              = 1121;
    generated_transaction_create_request                     = 1122;
    generated_transaction_create_response                    = 1123;
    generated_transaction_query_request                      = 1124;
    generated_transaction_query_response                     = 1125;
    generated_transaction_update_request                     = 1126;
    generated_transaction_update_response                    = 1127;
    generated_transaction_remove_request                     = 1128;
    generated_transaction_remove_response                    = 1129;
    producer_create_request                                  = 1130;
    producer_create_response                                 = 1131;
    producer_query_request                                   = 1132;
    producer_query_response                                  = 1133;
    producer_update_request                                  = 1134;
    producer_update_response                                 = 1135;
    producer_remove_request                                  = 1136;
    producer_remove_response                                 = 1137;
    account_control_history_create_request                   = 1138;
    account_control_history_create_response                  = 1139;
    account_control_history_query_request                    = 1140;
    account_control_history_query_response                   = 1141;
    account_control_history_update_request                   = 1142;
    account_control_history_update_response                  = 1143;
    account_control_history_remove_request                   = 1144;
    account_control_history_remove_response                  = 1145;
    public_key_history_create_request                        = 1146;
    public_key_history_create_response                       = 1147;
    public_key_history_query_request                         = 1148;
    public_key_history_query_response                        = 1149;
    public_key_history_update_request                        = 1150;
    public_key_history_update_response                       = 1151;
    public_key_history_remove_request                        = 1152;
    public_key_history_remove_response                       = 1153;
    table_id_create_request                                  = 1154;
    table_id_create_response                                 = 1155;
    table_id_query_request                                   = 1156;
    table_id_query_response                                  = 1157;
    table_id_update_request                                  = 1158;
    table_id_update_response                                 = 1159;
    table_id_remove_request                                  = 1160;
    table_id_remove_response                                 = 1161;
    resource_limits_create_request                           = 1162;
    resource_limits_create_response                          = 1163;
    resource_limits_query_request                            = 1164;
    resource_limits_query_response                           = 1165;
    resource_limits_update_request                           = 1166;
    resource_limits_update_response                          = 1167;
    resource_limits_remove_request                           = 1168;
    resource_limits_remove_response                          = 1169;
    resource_usage_create_request                            = 1170;
    resource_usage_create_response                           = 1171;
    resource_usage_query_request                             = 1172;
    resource_usage_query_response                            = 1173;
    resource_usage_update_request                            = 1174;
    resource_usage_update_response                           = 1175;
    resource_usage_remove_request                            = 1176;
    resource_usage_remove_response                           = 1177;
    resource_limits_state_create_request                     = 1178;
    resource_limits_state_create_response                    = 1179;
    resource_limits_state_query_request                      = 1180;
    resource_limits_state_query_response                     = 1181;
    resource_limits_state_update_request                     = 1182;
    resource_limits_state_update_response                    = 1183;
    resource_limits_state_remove_request                     = 1184;
    resource_limits_state_remove_response                    = 1185;
    resource_limits_config_create_request                    = 1186;
    resource_limits_config_create_response                   = 1187;
    resource_limits_config_query_request                     = 1188;
    resource_limits_config_query_response                    = 1189;
    resource_limits_config_update_request                    = 1190;
    resource_limits_config_update_response                   = 1191;
    resource_limits_config_remove_request                    = 1192;
    resource_limits_config_remove_response                   = 1193;
    account_history_create_request                           = 1194;
    account_history_create_response                          = 1195;
    account_history_query_request                            = 1196;
    account_history_query_response                           = 1197;
    account_history_update_request                           = 1198;
    account_history_update_response                          = 1199;
    account_history_remove_request                           = 1200;
    account_history_remove_response                          = 1201;
    action_history_create_request                            = 1202;
    action_history_create_response                           = 1203;
    action_history_query_request                             = 1204;
    action_history_query_response                            = 1205;
    action_history_update_request                            = 1206;
    action_history_update_response                           = 1207;
    action_history_remove_request                            = 1208;
    action_history_remove_response                           = 1209;
    reversible_block_create_request                          = 1210;
    reversible_block_create_response                         = 1211;
    reversible_block_query_request                           = 1212;
    reversible_block_query_response                          = 1213;
    reversible_block_update_request                          = 1214;
    reversible_block_update_response                         = 1215;
    reversible_block_remove_request                          = 1216;
    reversible_block_remove_response                         = 1217;



    permission_level_check_request                           = 2000;
    permission_level_check_response                          = 2001;
}

message error {
    optional uint32 code = 1;
    optional string msg  = 2;
}

message meta {
    optional uint32 id          = 1;
    optional uint32 service_id  = 2;
    optional uint32 method_id   = 3;
}

message rpc_request {
    optional meta        head    = 1;
    optional object_type type    = 2;
    optional message_id  msg_id  = 3;
    optional bytes       data    = 4;
}

message rpc_response {
    optional message_id  msg_id = 1;
    optional object_type type   = 2;
    optional error       err    = 3;
    optional bytes       data   = 4;
}

